src/main/java/de/unijena/cheminf/fragment/fingerprint/FragmentFingerprinter.java

```diff
@@ -0,0 +1,464 @@
+            if(tmpUniqueSmiles == null || aUniqueSmilesToFrequencyMap.get(tmpUniqueSmiles) == null) {
+                throw new NullPointerException("aUniqueSmilesToFrequencyMap (Map of string and integer instances) contains " +
+                        "instances that are null.");
+            }
+            if(tmpUniqueSmiles.isBlank() || tmpUniqueSmiles.isEmpty()) {
+                throw new IllegalArgumentException("aUniqueSmilesToFrequencyMap (Map of strings an integer instances) contains strings that are blank/empty.");
+            }
+            tmpListOfUniqueSmiles.add(tmpUniqueSmiles);
+        }
+       return this.createCountArray(tmpListOfUniqueSmiles);
+    }
+    //
+    /**
+     * Returns the count array for the specified list (molecule).
+     * @see #getCountArray(Map)
+     *
+     * @param aListOfUniqueSmiles is a list that stores fragments in the form of unique SMILES.
+     * @return int[] count array
+     * @throws NullPointerException is thrown if the list aListOfUniqueSmiles is null.
+     * @throws IllegalArgumentException is thrown if the list aListOfUniqueSmiles contains blank/empty strings.
+     */
+    public int[] getCountArray(List<String> aListOfUniqueSmiles) throws NullPointerException, IllegalArgumentException {
+        Objects.requireNonNull(aListOfUniqueSmiles, "aListOfUniqueSmiles (list of string instances) is null.");
+        for (String tmpUniqueSmiles : aListOfUniqueSmiles) {
+            Objects.requireNonNull(tmpUniqueSmiles, "aListOfUniqueSmiles (at least one list element) is null.");
+            if (tmpUniqueSmiles.isBlank() || tmpUniqueSmiles.isEmpty()) {
+                throw new IllegalArgumentException("aListOfUniqueSmiles (at least one list element) is blank/empty.");
+            }
+        }
+        Set<String> tmpSet = new HashSet<>(aListOfUniqueSmiles);
+        List<String> tmpListWithoutPossibleDuplicates = new ArrayList<>(tmpSet);
+        return this.createCountArray(tmpListWithoutPossibleDuplicates);
+    }
+    //
+    /**
+     * Generates count array for the specified list (molecule).
+     * Among other things, already generated results are used to generate the array.
+     * For example, if a count fingerprint has already been generated for the given list of unique SMILES or for
+     * the given molecule, the result of the count fingerprint is expanded into an array. Otherwise,
+     * the count fingerprint is generated first and then the count array.
+     *
+     * @param listOfUniqueSmiles is a list that stores fragments in the form of unique SMILES.
+     * @return int[] count array
+     */
+    private int[] createCountArray(List<String> listOfUniqueSmiles) {
+        int[] tmpCountArray = new int[this.fragmentArray.length];
+        if(this.cacheRawCountMap != null) {
+            Collections.sort(listOfUniqueSmiles);
+            Collections.sort(this.copyListCount);
+            if(listOfUniqueSmiles.equals(this.copyListCount)) {
```

Comment:
I have a few problems here: You are sorting, i.e. changing(!), the given input list. Unexpectedly modifying the variables given to you as parameters is bad style. Or are you sure that these are only temporary variables that your oublic method created and passed to this method? Also, did you check/test whether list.equals(list) does what you think it does?

Reply:
See (https://stackoverflow.com/questions/13501142/java-arraylist-how-can-i-tell-if-two-lists-are-equal-order-not-mattering  - Answer from Jacob Schoen):

You could sort both lists using Collections.sort() and then use the equals method. A slighly better solution is to first check if they are the same length before ordering, if they are not, then they are not equal, then sort, then use equals. For example if you had two lists of Strings it would be something like:

```
public  boolean equalLists(List<String> one, List<String> two){     
    if (one == null && two == null){
        return true;
    }

    if((one == null && two != null) 
      || one != null && two == null
      || one.size() != two.size()){
        return false;
    }

    //to avoid messing the order of the lists we will use a copy
    //as noted in comments by A. R. S.
    one = new ArrayList<String>(one); 
    two = new ArrayList<String>(two);   

    Collections.sort(one);
    Collections.sort(two);      
    return one.equals(two);
}
```

Reply:
Thanks, Felix!
But for this routine here, since it is only for checking whether the same input is presented again and we can return the cached fingerprint or not, I proposed on Wednesday to be more defensive here and just compare the lists in a basic way. In the unlikely case that a user sorts their list in some way between two calls to the method, well, bad luck, we will re-generate the fingerprint (which is also basically an instantaneous computation). So I guess we could use Felix' code block without the copying and sorting.

Reply:
Just my two cents for a general solution